Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -171,392 +177,470 @@ NonnullRefPtr<CSSStyleSheet> Parser::parse_as_stylesheet(TokenStream<T>& tokens)
         auto stream = TokenStream(selector_parts);
-        auto selector = parse_single_selector(stream);
+        auto selector = parse_complex_selector(stream, false);
         if (selector)
             selectors.append(selector.release_nonnull());
+        else
+            return {};
     }
 
-    return selectors;
-}
+    if (selectors.is_empty())
+        return {};
 
-NonnullRefPtrVector<Selector> Parser::parse_a_relative_selector()
-{
-    return parse_a_relative_selector(m_token_stream);
+    return selectors;
 }
 
 template<typename T>
-NonnullRefPtrVector<Selector> Parser::parse_a_relative_selector(TokenStream<T>& tokens)
+Optional<SelectorList> Parser::parse_a_relative_selector_list(TokenStream<T>& tokens)
 {
-    dbgln_if(CSS_PARSER_TRACE, "Parser::parse_a_relative_selector");
+    dbgln_if(CSS_PARSER_DEBUG, "Parser::parse_a_relative_selector_list");
 
     auto comma_separated_lists = parse_as_comma_separated_list_of_component_values(tokens);
 
     NonnullRefPtrVector<Selector> selectors;
-
     for (auto& selector_parts : comma_separated_lists) {
         auto stream = TokenStream(selector_parts);
-        auto selector = parse_single_selector(stream, true);
+        auto selector = parse_complex_selector(stream, true);
         if (selector)
             selectors.append(selector.release_nonnull());
+        else
+            return {};
     }
 
+    if (selectors.is_empty())
+        return {};
+
     return selectors;
 }
 
-template<typename T>
-RefPtr<Selector> Parser::parse_single_selector(TokenStream<T>& tokens, bool is_relative)
+RefPtr<Selector> Parser::parse_complex_selector(TokenStream<StyleComponentValueRule>& tokens, bool allow_starting_combinator)
 {
-    dbgln_if(CSS_PARSER_TRACE, "Parser::parse_single_selector");
+    dbgln_if(CSS_PARSER_DEBUG, "Parser::parse_complex_selector");
 
-    // FIXME: Bring this all in line with the spec. https://www.w3.org/TR/selectors-4/
+    Vector<Selector::CompoundSelector> compound_selectors;
 
-    Vector<Selector::ComplexSelector> selectors;
-
-    auto check_for_eof_or_whitespace = [&](T& current_value) -> bool {
-        if (current_value.is(Token::Type::EndOfFile))
-            return true;
+    auto first_selector = parse_compound_selector(tokens);
+    if (first_selector.is_error())
+        return {};
+    if (!allow_starting_combinator) {
+        if (first_selector.value().combinator != Selector::Combinator::Descendant)
+            return {};
+        first_selector.value().combinator = Selector::Combinator::None;
+    }
+    compound_selectors.append(first_selector.value());
 
-        if (current_value.is(Token::Type::Whitespace)) {
-            tokens.reconsume_current_input_token();
-            return true;
+    while (tokens.has_next_token()) {
+        auto compound_selector = parse_compound_selector(tokens);
+        if (compound_selector.is_error()) {
+            if (compound_selector.error() == SelectorParsingResult::Done)
```

Comment:
This sounds funny! We have an `Error` that actually signifies success.

Reply:
It's a bit weird yeah. It's the best option I could think of, as we have 3 states: "returning a value", "returning nothing because we're at the end" and "returning nothing because it's an error". There's probably a better way of doing that though!

Reply:
`Optional<Result<T>>`? I don't think that would be better though. Let's keep it like that but also keep the in mind that it's weird. Maybe we find some data structure that represents our use-case better some day.